### PR TITLE
add data integrity verifiable presentation as proof of possession

### PR DESCRIPTION
**Pull request by F-Node-Karlsruhe**
Added a linked data proof type for the proof of possession in the OID4VCI flow

Original PR: https://bitbucket.org/openid/connect/pull-requests/487/add-ldp_vp-as-proof-of-possession

### CODE REVIEW

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:804]:
```suggestion
The verifiable presentation as defined by [@VC_DATA_2.0] signed using Data Integrity that are used as Key Proofs MUST contain the following elements:
```

saying just 'verifiable presentation' unfortunately does not differentiate enough from key proof type JWT

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:738]:
```suggestion
* `ldp_vp_2.0`: A verifiable presentation signed using the data integrity proof defined in [@VC_DATA_2.0], and where the proof of possession MUST be done in accordance with [@DI]. When `proof_type` is set to `ldp_vp_2.0`, the `proof` object MUST include a `ldp_vp` claim containing a [verifiable presentation](https://www.w3.org/TR/vc-data-model-2.0/#presentations-0) defined in (#ldp_vp-proof-type).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:804]: Is Data Integrity a defined term? If yes, please capitalize this in the entire PR/Document

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:811]:
```suggestion
      * `challenge`: REQUIRED (string). The value type of this claim MUST be a string, where the value is a server-provided `c_nonce`. It MUST be present when the Wallet received server-provided `c_nonce`.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:813]:
```suggestion
The Credential Issuer MUST validate that the `proof` is actually signed with a key in possession of the Holder.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:1897]:
```suggestion
For better suiting the W3C Data Integrity Proof ecosystem the following non-normative example displays a Credential request with the proof of possession type set to `ldp_vp`, using a verifiable presentation.
```

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md:804]: Was just one more occurrence

https://github.com/openid/OpenID4VCI/pull/66/commits/93c20c80e547ce70b655ca15d38477b7e7f87200

**Review by peppelinux — APPROVED** [review]

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (decentralgabe) [inline openid-4-verifiable-credential-issuance-1_0.md:738]: you link to the V2 data model down here but above your example shows a v1 VC, you may want to update the above example to point to the v2 context and use v2 properties

Comment (decentralgabe) [reply openid-4-verifiable-credential-issuance-1_0.md:738]: since there has been a rename from `linked data proofs` to `data integrity proofs` should this property be `di_vp_2.0`?

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (decentralgabe) [inline openid-4-verifiable-credential-issuance-1_0.md:807]: does not need to be a DID - just the verificationMethod identifier

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md:738]: I aligned to the spec, where Data Integrity Proofs are still named with Linked Data Proofs `ldp`. I am open to change that throughout the spec. But for now (the PR) i thought it would be better to stick to the used naming, as it should not touch the general naming imo.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md]: That's right. I changed that.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (decentralgabe) [inline openid-4-verifiable-credential-issuance-1_0.md:823]:
```suggestion
         "@context": [
             "https://www.w3.org/ns/credentials/v2",
             "https://www.w3.org/ns/credentials/examples/v2"
         ],
```

File: examples/credential_request_ldp_vc_vp.json
Comment (decentralgabe) [inline examples/credential_request_ldp_vc_vp.json:23]:
```suggestion
         "@context": [
             "https://www.w3.org/ns/credentials/v2",
             "https://www.w3.org/ns/credentials/examples/v2"
         ],
```

**Comment by F-Node-Karlsruhe** [comment]
For the use case of proof of possession i actually advocate removing the version number 2.0 `ldp_vp_2.0`-> `ldp_vp` again as the proof already carries this information itself and it only makes the spec more complicated.

One verifying a ldp_vp should both be capable of verifying a 1.0 as well as a 2.0 depending on the referenced context. Especially bearing in mind that the data model only changed slightly and the general form of the presentation remains the same. Most libraries already are capable of treating both versions at the same time.

**Comment by Sakurann** [comment]
@dlongley does the mechanism described in the PR looks good to you? would appreciate an approval/request for changes.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (dlongley) [inline openid-4-verifiable-credential-issuance-1_0.md:807]: Minor tweak here:

```suggestion
      * `holder`: OPTIONAL. MUST be equivalent to the controller identifier (e.g. DID) for the verificationMethod identified by the `proof.verificationMethod` property.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (dlongley) [inline openid-4-verifiable-credential-issuance-1_0.md:813]: I don't know what to do about this language since it's "the right language / commonly used language" when considering other specifications around "proof of possession". However, this is technically incorrect for more or less all specs that use that terminology (so, not specific to this PR).

In short, "proof of possession" is really "proof of use", as the person who provides this proof does not actually need to "possess" the key / its secret material, they only need to be able to generate a signature over what is required (e.g., the VP above and the challenge, provided via `c_nonce` in OID4VCI). Someone who exposes an authz-protected API to use a particular key could grant anyone they wanted access to that API without ever exposing the private key material -- and a generated proof would therefore not be a "proof of possession" for such a user of this API.

I don't know that we want to take on trying to fix this language here in this one spot since every other "proof of possession" spec has a similar problem. But the Credential Issuer is *not capable* of performing the validation based on the language here -- and the same applies to other language used in other specs around proof of possession. If we want this to be the first place that we try to improve this to better communicate actual guarantees and requirements, here's a suggestion:

```suggestion
The Credential Issuer MUST verify that the `proof` was generated by a key associated with its verification method, thereby demonstrating that the Holder was capable of causing this key to be used to generate the proof.
```

**Review by dlongley — APPROVED** [review]
Approving with a few comments, one a minor clean up and the other a comment on language used in specs (generally) around "proof of possession" that promises / requires more than is technically possible.

**Comment by dlongley** [comment]
@Sakurann,

> does the mechanism described in the PR looks good to you? would appreciate an approval/request for changes.

Yes, approved -- with a couple of minor suggestions. This would be very helpful, thanks!

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md:813]: Even if it is different in the rest of the spec, i prefer your more precise description here. As you said: The proof of possession is actually a proof of being able to use/access the keys.

But this is rather up to @Sakurann or @tlodderstedt to decide.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:1901]: This is a section about credential offer, but example is about credential request. Please put this example in `#### Credential Request {#credential_request_ldp_vc}` section.

**Review by Sakurann — CHANGES_REQUESTED** [review]
example needs to be moved to an appropriate section

**Comment by Sakurann** [comment]
> +1 to keeping ldp_vp for simplicity.

why not  `di_vp`? shouldn't the use of a term `linked data proof` be deprecated as it is now called `data integrity`?

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:813]: yes, i would prefer not to make this change in this PR. feel free to raise an issue and address it more systematically if agreed to.

**Comment by dlongley** [comment]
> > +1 to keeping ldp_vp for simplicity.
> 
> why not `di_vp`? shouldn't the use of a term `linked data proof` be deprecated as it is now called `data integrity`?

My guess is that isn't not worth the churn to existing implementations or to have an inconsistency between `ldp_vc` and `di_vp` (vs. `ldp_vp`). Documentation could clarify the naming alias / history. Otherwise, if the thought is that it's still early enough to change -- then all `ldp_*` should be changed to `di_*` for consistency's sake with backwards compatibility for some time.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md]: https://github.com/openid/OpenID4VCI/pull/66/commits/31051475939b42d2cb2e93611f6030d15522e7b8

**Comment by F-Node-Karlsruhe** [comment]
> > +1 to keeping ldp_vp for simplicity.
> 
> why not `di_vp`? shouldn't the use of a term `linked data proof` be deprecated as it is now called `data integrity`?

We had this thought on this PR on bitbucket before and chose to stay consistent with the current spec (`ldp_*` for data integrity proofs) for this PR and move that change to another PR in the future. Pretty much how @dlongley outlined it.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:772]: sorry, I worry `[@DI]` is not clear enough. Can we please extend to `[@Data_Integrity]` everywhere? thanks.

```suggestion
* `ldp_vp`: A verifiable presentation signed using the Data Integrity Proof defined in [@VC_DATA_2.0] or [@VC_DATA], and where the proof of possession MUST be done in accordance with [@Data_Integrity]. When `proof_type` is set to `ldp_vp`, the `proof` object MUST include a `ldp_vp` claim containing a [verifiable presentation](https://www.w3.org/TR/vc-data-model-2.0/#presentations-0) defined in (#ldp_vp-proof-type).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:838]:
```suggestion
When a W3C Verifiable Presentation as defined by [@VC_DATA_2.0] or [@VC_DATA] signed using Data Integrity is used as Key Proof, it MUST contain the following elements:
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:840]:
```suggestion
  * in the payload of a W3C Verifiable Presentation itself,
```
not sure `payload` is most appropriate, but something along these lines

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:841]:
```suggestion
      * `holder`: OPTIONAL. MUST be equivalent to the controller identifier (e.g. DID) for the `verificationMethod` value identified by the `proof.verificationMethod` property.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:845]: Is `challenge` Optional when `c_nonce` is not provided by the issuer? if so, a language like `REQUIRED when ... . MUST NOT be used otherwise.` is probably more appropriate.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:772]: on another note, i think it is important to clarify the value of `ldp_vc` claim is an object, since when proof type is jwt or cwt, the value is a string.
```suggestion
* `ldp_vp`: An object representing a W3C Verifiable Presentation signed using the Data Integrity Proof as defined in [@VC_DATA_2.0] or [@VC_DATA], and where the proof of possession MUST be done in accordance with [@DI]. When `proof_type` is set to `ldp_vp`, the `proof` object MUST include a `ldp_vp` claim containing a [W3C Verifiable Presentation](https://www.w3.org/TR/vc-data-model-2.0/#presentations-0) defined in (#ldp_vp-proof-type).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:843]: I am not sure the structure is accurate. proof is part of a w3c verifiable presentation, right? so isn't the following more accurate?
```
When a W3C Verifiable Presentation as defined by [@VC_DATA_2.0] or [@VC_DATA] signed using Data Integrity is used as Key Proof, it MUST contain the following elements:

* holder: ..
* proof: ...
    *  domain
    * challenge
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:1963]: I would suggest removing an example in lines 1956~1958.

```suggestion
The following is a non-normative example of a Credential request with the key proof type `ldp_vp`:

<{{examples/credential_request_ldp_vc_vp.json}}

```

**Review by Sakurann — CHANGES_REQUESTED** [review]
concrete changes suggested

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md]: https://github.com/openid/OpenID4VCI/pull/66/commits/e6822532a4d7d2bbc1313376ac470fe6348d6a21

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md]: https://github.com/openid/OpenID4VCI/pull/66/commits/9e31825398e426fb22da4c09c7259e0647dcb677

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (F-Node-Karlsruhe) [reply openid-4-verifiable-credential-issuance-1_0.md]: https://github.com/openid/OpenID4VCI/pull/66/commits/9e31825398e426fb22da4c09c7259e0647dcb677

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:844]:
```suggestion
      * `challenge`: REQUIRED when the Credential Issuer has provided a `c_nonce`. MUST NOT be used otherwise. String, where the value is a server-provided `c_nonce`. It MUST be present when the Wallet received server-provided `c_nonce`.
```

**Review by Sakurann — APPROVED** [review]
looks good. approving assuming my one suggestion will be accepted

**Review by tlodderstedt — APPROVED** [review]

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:1517]: resolving compiling errors
```suggestion
```

Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:1517]:
```suggestion
```

Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:1517]: @F-Node-Karlsruhe can you please accept this change, which I think will allow for the md to compile? I wanted to merge, but the file was not compiling correctly.

**Comment by Sakurann** [comment]
@F-Node-Karlsruhe thank you for the PR and working with us. apologies it took the wg a while to merge this!

**Comment by Sakurann** [comment]
@F-Node-Karlsruhe what name can I use to add you to the contributors in the acknowledgements section of the specificaiton?

**Comment by F-Node-Karlsruhe** [comment]
> @F-Node-Karlsruhe what name can I use to add you to the contributors in the acknowledgements section of the specificaiton?

Christian Fries - European EPC Competence Center

https://www.researchgate.net/profile/Christian_Fries4

thank you for mentioning me :)

**Pull request merged**
